File: 235.lowest-common-ancestor-of-a-binary-search-tree/solution.kt
/**
 * Definition for a binary tree node.
 * class TreeNode(var `val`: Int = 0) {
 *     var left: TreeNode? = null
 *     var right: TreeNode? = null
 * }
 */

class Solution {
    fun lowestCommonAncestor(root: TreeNode?, p: TreeNode?, q: TreeNode?): TreeNode? {
        val lo = minOf(p!!.`val`, q!!.`val`)
        val hi = maxOf(p!!.`val`, q!!.`val`)
        var node = root
        
        while (node != null) {
            node = when {
                node.`val` > hi -> node.left
                node.`val` < lo -> node.right
                else -> return node
            }
        } 
        return node
    }
}

File: 235.lowest-common-ancestor-of-a-binary-search-tree/solution.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val   int
 *     Left  *TreeNode
 *     Right *TreeNode
 * }
 */

func lowestCommonAncestor(root, p, q *TreeNode) *TreeNode {
    var lo, hi int
    if p.Val < q.Val {
        lo = p.Val
        hi = q.Val
    } else {
        lo = q.Val
        hi = p.Val
    }
    for root != nil {
        if root.Val < lo {
            root = root.Right
        } else if root.Val > hi {
            root = root.Left
        } else {
            return root
        }
    }
    return nil
}